server/src/main/java/org/elasticsearch/monitor/metrics/IndicesMetrics.java

```diff
@@ -0,0 +1,169 @@
+                    "unit",
+                    () -> new LongWithAttributes(cache.getOrRefresh().get(indexMode).numDocs)
+                )
+            );
+            metrics.add(
+                registry.registerLongGauge(
+                    "es.indices." + name + ".bytes.total",
+                    "total size in bytes of " + name + " indices",
+                    "unit",
+                    () -> new LongWithAttributes(cache.getOrRefresh().get(indexMode).numBytes)
+                )
+            );
+        }
+        return metrics;
+    }
+
+    @Override
+    protected void doStart() {
+        metrics.addAll(registerAsyncMetrics(registry, stateCache));
+    }
+
+    @Override
+    protected void doStop() {
+        stateCache.stopRefreshing();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+        metrics.forEach(metric -> {
+            try {
+                metric.close();
+            } catch (Exception e) {
+                logger.warn("metrics close() method should not throw Exception", e);
+            }
+        });
+    }
+
+    static class IndexStats {
+        int numIndices = 0;
+        long numDocs = 0;
+        long numBytes = 0;
+    }
+
+    private static class IndicesStatsCache extends SingleObjectCache<Map<IndexMode, IndexStats>> {
+        private static final Map<IndexMode, IndexStats> MISSING_STATS;
+        static {
+            MISSING_STATS = new EnumMap<>(IndexMode.class);
+            for (IndexMode value : IndexMode.values()) {
+                MISSING_STATS.put(value, new IndexStats());
+            }
+        }
+
+        private boolean refresh;
+        private final IndicesService indicesService;
+
+        IndicesStatsCache(IndicesService indicesService, TimeValue interval) {
+            super(interval, MISSING_STATS);
+            this.indicesService = indicesService;
+            this.refresh = true;
+        }
+
+        private Map<IndexMode, IndexStats> internalGetIndicesStats() {
+            Map<IndexMode, IndexStats> stats = new EnumMap<>(IndexMode.class);
+            for (IndexMode mode : IndexMode.values()) {
+                stats.put(mode, new IndexStats());
+            }
+            for (IndexService indexService : indicesService) {
+                for (IndexShard indexShard : indexService) {
+                    if (indexShard.isSystem()) {
+                        continue; // skip system indices
+                    }
+                    final IndexMode indexMode = indexShard.indexSettings().getMode();
+                    final IndexStats indexStats = stats.get(indexMode);
+                    final ShardRouting shardRouting = indexShard.routingEntry();
+                    if (shardRouting.primary() && shardRouting.shardId().id() == 0) {
+                        indexStats.numIndices++;
+                    }
+                    try {
+                        indexStats.numDocs += indexShard.commitStats().getNumDocs();
```

Comment:
should we only keep track of these stats for primary shards?

Reply:
This works for me as well. If we decide to use only primaries, I think we should exclude relocating shards.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/113737/commits/44456fa0d024eb641a230a45b2a388e77f269f41

Reply:
👍